### PR TITLE
Add draft of spec

### PR DESCRIPTION
**Pull request by mreichhoff**
This is quite rough, but is a start.

### CODE REVIEW

File: requestStorageAccessForOrigin.bs [hunk 1]
```diff
@@ -0,0 +1,222 @@
...
+  <dt>[=powerful feature/permission query algorithm=]</dt>
+  <dd>
+    To query the "<a permission><code>top-level-storage-access</code></a>" permission, given a {{PermissionDescriptor}} |permissionDesc| and a {{PermissionStatus}} |status|, run the following steps:
+
+    1. Set |status|'s {{PermissionStatus/state}} to |permissionDesc|'s [=permission state=].
+    1. If |status|'s {{PermissionStatus/state}} is [=permission/denied=], set |status|'s {{PermissionStatus/state}} to [=permission/prompt=].
+
+        Note: The "denied" permission state is not revealed to avoid exposing the user's decision to developers. This is done to prevent retaliation against the user and repeated prompting to the detriment of the user experience.
+  </dd>
+  <dt>[=powerful feature/permission key generation algorithm=]</dt>
```
Comment (johannhof) [inline requestStorageAccessForOrigin.bs:179]: As per the latest update of Permissions this also needs to specify a [permission key type](https://w3c.github.io/permissions/#dfn-permission-key-type)

File: requestStorageAccessForOrigin.bs [hunk 2]
```diff
@@ -0,0 +1,222 @@
...
+        1. If |current origin| and |top-level site| are not [=same site=] with one another and |current origin| and |embedded origin| are not [=same origin=] with one another, return false.
+1. Let |descriptor| be a newly created {{TopLevelStorageAccessPermissionDescriptor}} with <code>name</code> set to "<a permission><code>top-level-storage-access</code></a>" and with <code>requestedOrigin</code> set to |embedded origin|.
+1. Let |existing state| be the result of checking |descriptor|'s [=permission state=] with |settings|.
+1. If |existing state| is "granted", return true.
+1. Return false.
+
+To <dfn type="abstract-op">determine the top-level storage access policy</dfn> for {{TopLevelStorageAccessPermissionDescriptor}} |descriptor|, with {{Document}} |doc| and {{Promise}} |p|, run these steps:
+1. Let |settings| be |doc|'s [=relevant settings object=].
+<!-- TODO: where to check this, exactly? -->
+1. Let |existing state| be the result of checking |descriptor|'s [=permission state=] with |settings|.
```
Comment (johannhof) [inline requestStorageAccessForOrigin.bs:140]:
```suggestion
1. Let |existing state| be |descriptor|'s [=permission state=] with |settings|.
```

**Review by johannhof — APPROVED** [review]
I think this is fine to merge as WIP state, though there are still a bunch of things to improve :)

I had a few suggestions, most of them can be deferred to follow-up issues depending on how you prefer to iterate on this spec.

File: requestStorageAccessForOrigin.bs [hunk 3]
```diff
@@ -0,0 +1,222 @@
...
+spec:html; type:dfn; for:site; text:same site
+spec:webidl; type:dfn; text:resolve
+</pre>
+
+<pre class="anchors">
+urlPrefix: https://tc39.github.io/ecma262/; spec: ECMASCRIPT
+    text: agent cluster; url: #sec-agent-clusters; type: dfn
+urlPrefix: https://infra.spec.whatwg.org/; spec: INFRA
+    text: implementation-defined; url: #implementation-defined; type: dfn
+urlPrefix: https://w3c.github.io/webdriver/webdriver-spec.html#; spec: webdriver
```
[the remark this answers is not in the document]
Comment (mreichhoff) [reply requestStorageAccessForOrigin.bs:28]: yup, this just got missed after I pulled out the webdriver stuff. Removing.

File: requestStorageAccessForOrigin.bs [hunk 4]
```diff
@@ -0,0 +1,222 @@
...
+Editor: Matt Reichhoff, w3cid 138889, Google https://google.com, mreichhoff@google.com
+Editor: Johann Hofmann, w3cid 120436, Google https://google.com, johannhof@google.com
+Level: None
+Markup Shorthands: markdown yes, css no
+Complain About: accidental-2119 true
+</pre>
+
+<!-- TODO: evaluate which is needed -->
+<pre class=link-defaults>
+spec:html; type:dfn; for:site; text:same site
```
[the remark this answers is not in the document]
Comment (mreichhoff) [reply requestStorageAccessForOrigin.bs:19]: one is needed, or it'll complain about the reference to same site being ambiguous. I think this one is what rSA uses, so I'll follow suit.

**Comment by mreichhoff** [comment]
closing in favor of a version that requires more-explicit opt-in via rSA call.

**Pull request closed without merging**